cps/kni.c

```diff
@@ -55,6 +62,240 @@ kni_change_mtu(uint8_t port_id, unsigned new_mtu)
+}
+
+static int
+modify_link(struct mnl_socket *nl, struct rte_kni *kni,
+	const char *kni_name, int if_up)
+{
+	char buf[MNL_SOCKET_BUFFER_SIZE];
+	struct nlmsghdr *nlh;
+	struct ifinfomsg *ifm;
+	unsigned int seq, flags = 0;
+	int pid;
+	int ret = 0;
+
+	if (if_up)
+		flags |= IFF_UP;
+	else
+		flags &= ~IFF_UP;
+
+	nlh = mnl_nlmsg_put_header(buf);
+	nlh->nlmsg_type	= RTM_NEWLINK;
+	nlh->nlmsg_flags = NLM_F_REQUEST|NLM_F_ACK;
+	nlh->nlmsg_seq = seq = time(NULL);
+
+	ifm = mnl_nlmsg_put_extra_header(nlh, sizeof(*ifm));
+	ifm->ifi_family = AF_UNSPEC;
+	ifm->ifi_change = IFF_UP;
+	ifm->ifi_flags = flags;
+
+	mnl_attr_put_str(nlh, IFLA_IFNAME, kni_name);
+
+	/*
+	 * The KNI library registers callbacks for MTU changes and
+	 * interface up/down events through ethtool. When these
+	 * operations are requested to the kernel through netlink
+	 * messages, they go back to DPDK in userspace when
+	 * rte_kni_handle_request() is invoked. Therefore, if we
+	 * issue this request, we need another process to call
+	 * rte_kni_handle_request() to allow it to proceed.
+	 *
+	 * The DPDK documentation suggests using ifconfig from a
+	 * shell to bring a KNI up after the application starts and
+	 * calls rte_kni_handle_request(). To do so automatically,
+	 * we have to fork a child process to issue the request
+	 * while the parent calls rte_kni_handle_request().
+	 */
+
+	pid = fork();
+	if (pid == -1) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"cps: fork failed, can't modify KNI %s link: %s\n",
+			kni_name, strerror(errno));
+		return -1;
+	} else if (pid == 0) {
+		/*
+		 * Send request to kernel, which will be sent back
+		 * to userspace for the parent to handle.
+		 */
+		int ret = mnl_socket_sendto(nl, nlh, nlh->nlmsg_len);
+		if (ret < 0) {
+			RTE_LOG(ERR, GATEKEEPER,
+				"cps: mnl_socket_sendto: cannot bring KNI %s %s: %s\n",
+				kni_name, if_up ? "up" : "down",
+				strerror(errno));
+			_exit(EXIT_FAILURE);
+		}
+		_exit(EXIT_SUCCESS);
+	} else {
+		unsigned int attempts = 0;
+		unsigned int portid = mnl_socket_get_portid(nl);
+		do {
+			int status;
+
+			/* Try to process child's request. */
+			int ret = rte_kni_handle_request(kni);
+			if (ret < 0) {
+				RTE_LOG(ERR, KNI, "%s: error in handling userspace request\n",
+					__func__);
+				goto next;
+			}
+
+			/* Check if child has finished submitting request. */
+			ret = waitpid(pid, &status, WNOHANG);
+			if (ret == 0) {
+				/* Keep trying to handle the KNI request. */
+				goto next;
+			} else if (ret == -1) {
+				RTE_LOG(ERR, GATEKEEPER, "cps: waitpid: %s\n",
+					strerror(errno));
+				goto kill;
+			}
+
+			/* Verify interface configuration message was valid. */
+			ret = mnl_socket_recvfrom(nl, buf, sizeof(buf));
+			if (ret == -1) {
+				RTE_LOG(ERR, GATEKEEPER, "cps: mnl_socket_recvfrom: cannot bring KNI %s %s: %s\n",
+					kni_name, if_up ? "up" : "down",
+					strerror(errno));
+				return ret;
+			}
+
+			ret = mnl_cb_run(buf, ret, seq, portid, NULL, NULL);
```

Comment:
Could you add a comment explaining what `mnl_cb_run()` does? I tried to figure out reading the following link, but the description is too short to make sense:

https://www.netfilter.org/projects/libmnl/doxygen/group__callback.html